pkg/maps/lbmap/lbmap.go

```diff
@@ -130,26 +132,51 @@ func (lbmap *LBBPFMap) UpsertService(p *UpsertServiceParams) error {
 		return fmt.Errorf("Unable to update reverse NAT %+v => %+v: %s", revNATKey, revNATValue, err)
 	}
 
-	if err := updateMasterService(svcKey, len(backendIDs), int(p.ID), p.Type, p.Local,
+	if err := updateMasterService(svcKey, backendLen, int(p.ID), p.Type, p.Local, p.NatPolicy,
 		p.SessionAffinity, p.SessionAffinityTimeoutSec, p.CheckSourceRange); err != nil {
-
 		deleteRevNatLocked(revNATKey)
 		return fmt.Errorf("Unable to update service %+v: %s", svcKey, err)
 	}
 
-	for i := slot; i <= p.PrevActiveBackendCount; i++ {
-		svcKey.SetBackendSlot(i)
-		if err := deleteServiceLocked(svcKey); err != nil {
-			log.WithFields(logrus.Fields{
-				logfields.ServiceKey:  svcKey,
-				logfields.BackendSlot: svcKey.GetBackendSlot(),
-			}).WithError(err).Warn("Unable to delete service entry from BPF map")
+	if backendsOk {
+		for i := slot; i <= p.PrevActiveBackendCount; i++ {
+			svcKey.SetBackendSlot(i)
+			if err := deleteServiceLocked(svcKey); err != nil {
+				log.WithFields(logrus.Fields{
+					logfields.ServiceKey:  svcKey,
+					logfields.BackendSlot: svcKey.GetBackendSlot(),
+				}).WithError(err).Warn("Unable to delete service entry from BPF map")
+			}
 		}
 	}
 
 	return nil
 }
 
+// UpsertService inserts or updates the given service in a BPF map.
+//
+// The corresponding backend entries (identified with the given backendIDs)
+// have to exist before calling the function.
+//
+// The service's prevActiveBackendCount denotes the count of previously active
+// backend entries that were added to the BPF map so that the function can remove
+// obsolete ones.
+func (lbmap *LBBPFMap) UpsertService(p *UpsertServiceParams) error {
+	if p.ID == 0 {
+		return fmt.Errorf("Invalid svc ID 0")
+	}
+	if err := lbmap.upsertServiceProto(p,
+		p.IPv6 || p.NatPolicy == loadbalancer.SVCNatPolicyNat46); err != nil {
+		return err
+	}
+	if p.NatPolicy == loadbalancer.SVCNatPolicyNat46 {
+		if err := lbmap.upsertServiceProto(p, false); err != nil {
+			return err
```

Comment:
hi @borkmann - rather late to the party, but need one silly clarification. For a NAT46 service, why do we create an additional Maglev table here for the IPv4 dummy frontend? I thought the backend selection for NAT46 happens in the IPv6 codepath...?

Reply:
@julianwiedmann That's a good question. I will defer to @borkmann for an answer, but to me it seems a bug. Also, that should explain the failures in the L4LB GH action https://github.com/cilium/cilium/pull/20625.

Reply:
We shouldn't, the IPv4 to IPv6 has only dummy IPv4 service entry (0 backends) given it will recircle the packet into the IPv6 path. @julianwiedmann do we create a full-blown table?

Reply:
This one seems to be missing:
```
diff --git a/pkg/maps/lbmap/lbmap.go b/pkg/maps/lbmap/lbmap.go
index db6b83521e..6b1f68c402 100644
--- a/pkg/maps/lbmap/lbmap.go
+++ b/pkg/maps/lbmap/lbmap.go
@@ -87,15 +87,14 @@ func (lbmap *LBBPFMap) upsertServiceProto(p *datapathTypes.UpsertServiceParams,
                backends = p.PreferredBackends
        }
 
-       if p.UseMaglev && len(backends) != 0 {
-               if err := lbmap.UpsertMaglevLookupTable(p.ID, backends, ipv6); err != nil {
-                       return err
+       if backendsOk {
+               if p.UseMaglev && len(backends) != 0 {
+                       if err := lbmap.UpsertMaglevLookupTable(p.ID, backends, ipv6); err != nil {
+                               return err
+                       }
                }
-       }
 
-       if backendsOk {
                backendIDs := p.GetOrderedBackends()
-
                for _, backendID := range backendIDs {
                        if backendID == 0 {
                                return fmt.Errorf("Invalid backend ID 0")
```

Reply:
(Pushed here https://github.com/cilium/cilium/pull/20648)